ob1k-cache/src/main/java/com/outbrain/ob1k/cache/metrics/MonitoredTypedCache.java

```diff
@@ -0,0 +1,119 @@
+package com.outbrain.ob1k.cache.metrics;
+
+import java.util.Map;
+import com.google.common.base.Preconditions;
+import com.outbrain.ob1k.cache.EntryMapper;
+import com.outbrain.ob1k.cache.TypedCache;
+import com.outbrain.ob1k.concurrent.ComposableFuture;
+import com.outbrain.ob1k.concurrent.Consumer;
+import com.outbrain.ob1k.concurrent.Try;
+import com.outbrain.swinfra.metrics.api.MetricFactory;
+import com.outbrain.swinfra.metrics.api.Timer;
+
+/**
+ * A decorator adding metric monitoring to TypedCache.
+ *
+ * @author hunchback
+ */
+public class MonitoredTypedCache<K, V> implements TypedCache<K, V> {
+  protected final TypedCache<K, V> inner;
+
+  private final AsyncGetMetrics getAsyncMetrics;
+  private final AsyncBulkGetMetrics getBulkAsyncMetric;
+  private final AsyncOperationMetrics setAsyncMetrics;
+  private final AsyncOperationMetrics setBulkAsyncMetrics;
+  private final AsyncOperationMetrics deleteAsyncMetrics;
+
+  public MonitoredTypedCache(final TypedCache<K, V> inner, String cacheName, final MetricFactory metricFactory) {
+    Preconditions.checkNotNull(inner);
+    this.inner = inner;
+    Preconditions.checkNotNull(metricFactory);
+    final String component = inner.getClass().getSimpleName() + "." + cacheName;
+
+    getAsyncMetrics = new AsyncGetMetrics(metricFactory, component + ".getAsync");
```

Comment:
>Honestly, I don't like these little metrics classes in this context. I think a better abstraction might be to create consumer classes for each operation which you can also create once. I suspect it will also simplify the code.

By using Consumers you propose to split finished(...) functionality and attributes (hit/total,etc.) this does have merit in the case of get/bulk (sharing data but not logic) but has less merit in the case of AsyncOperationMetrics ... I would feel better addressing all Metrics in a similar way so would not be happy to treat get/bulk in a totally different way to how are deal with the basic metrics which reside within AsyncOperationMetrics.